# Выпускной проект по SQL

## Описание проекта

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.


## Цели исследования

Проанализировать базу данных и ответить на следующие вопросы:

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

### Подключение к базе и исследование таблиц

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

**Выведем первые строки таблиц:**

In [2]:
# чтобы выполнить SQL-запрос, используем Pandas
# выведем на экран таблицу books
query = '''SELECT * FROM books'''

con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [3]:
# выведем на экран таблицу authors
query = '''SELECT * FROM authors'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [4]:
# выведем на экран таблицу publishers
query = '''SELECT * FROM publishers'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

In [5]:
# выведем на экран таблицу ratings
query = '''SELECT * FROM ratings '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [6]:
# выведем на экран таблицу reviews
query = '''SELECT * FROM reviews'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

## Выполнение запросов

### Посчитайте, сколько книг вышло после 1 января 2000 года

In [7]:
query = '''SELECT COUNT(book_id) AS books_cnt
            FROM books 
            WHERE publication_date > '2000-01-01' '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

books_cnt
0        819

**Вывод: После 1 января 2020 года (не включая эту дату) вышло 819 книг.**

### Для каждой книги посчитайте количество обзоров и среднюю оценку;

*Будем использовать тип слиятния LEFT OUTER JOIN, т.к. нам нужно, чтобы в результат слияния вошли все записи из левой таблицы (books)*

In [12]:
query = '''SELECT b.book_id,
            b.title,
            COUNT(DISTINCT review_id) AS review_cnt,
            ROUND(AVG(rating),1) AS avg_rating

FROM books AS b
LEFT OUTER JOIN reviews AS rw ON b.book_id = rw.book_id 
LEFT OUTER JOIN ratings AS rt ON rt.book_id = b.book_id
GROUP by b.book_id
ORDER by avg_rating DESC '''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id                                              title  review_cnt  \
0        169  Crucial Conversations: Tools for Talking When ...           2   
1        901  The Walking Dead  Book One (The Walking Dead #...           2   
2         17                      A Dirty Job (Grim Reaper  #1)           4   
3        275                                          Geek Love           2   
4        321                                Homage to Catalonia           2   
..       ...                                                ...         ...   
995      772                             The Kitchen God's Wife           3   
996      316                  His Excellency: George Washington           2   
997      202                                      Drowning Ruth           3   
998      371                                              Junky           2   
999      303                               Harvesting the Heart           2   

     avg_rating  
0           5.0  
1           5.0  
2           5.0  
3           5.0  
4           5.0  
..          ...  
995         2.3  
996         2.0  
997         2.0  
998         2.0  
999         1.5  

[1000 rows x 4 columns]

<div class="alert alert-info"> <b>Комментарий студента:</b> Новый вывод:  для заданной базы с 1000 книг имеем средний рейтинг от 5.0 до 1.5 и количество обзоров от 7 (при задании сортировки ORDER by review_cnt DESC) до 0.</div>    

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [13]:
query = '''SELECT publisher,
            COUNT(book_id) AS books_cnt
FROM books AS b
LEFT OUTER JOIN publishers AS p ON b.publisher_id = p.publisher_id 
WHERE num_pages > 50 
GROUP by publisher
ORDER by books_cnt DESC
LIMIT 1'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher  books_cnt
0  Penguin Books         42

**Вывод: Больше всего книг толще 50 страниц выпустило издательство Penguin Books - 42 книги**

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [15]:
query = '''SELECT a.author,
          ROUND(AVG(rating),1) AS avg_rating
          
FROM books AS b
JOIN authors AS a ON a.author_id = b.author_id 
JOIN ratings AS rt ON rt.book_id = b.book_id
GROUP BY a.author
HAVING COUNT(rating_id) >50
ORDER BY avg_rating DESC
LIMIT 1'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author  avg_rating
0  J.K. Rowling/Mary GrandPré         4.3

<div class="alert alert-info"> <b>Комментарий студента:</b> Новый вывод:  Автор с самой высокой средней оценкой книг при условии что количество оценок больше 50 - J.K. Rowling/Mary GrandPre. </div>    

###  Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [17]:
# выведем пользователей, которые поставили больше 48 оценок
query = '''SELECT username,
        COUNT(rating_id) as rating_cnt
FROM ratings 
GROUP BY username
HAVING COUNT(rating_id) > 48
ORDER BY rating_cnt DESC'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

username  rating_cnt
0           paul88          56
1       martinadam          56
2      sfitzgerald          55
3        richard89          55
4   jennifermiller          53
5           xdavis          51
6   shermannatalie          50
7      lesliegibbs          50
8   vanessagardner          50
9          vcarter          49
10        walter49          49
11     andreaeaton          49
12         susan85          49

In [18]:
# посчитаем количество обзоров от пользователей, которые поставили больше 48 оценок
query = '''WITH 
        i as (SELECT username,
        COUNT(rating_id) as rating_cnt
FROM ratings 
GROUP BY username
HAVING COUNT(rating_id) > 48
ORDER BY rating_cnt DESC)

SELECT r.username,
        COUNT(review_id) AS review_cnt
FROM i 
INNER JOIN reviews as r on i.username=r.username
GROUP BY r.username'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

username  review_cnt
0         walter49          23
1      sfitzgerald          28
2   jennifermiller          25
3      andreaeaton          22
4           xdavis          18
5          vcarter          20
6           paul88          22
7      lesliegibbs          26
8          susan85          29
9       martinadam          27
10  vanessagardner          22
11       richard89          26
12  shermannatalie          24

In [19]:
#посчитаем среднее количество обзоров по пользователям из таблицы выше
query = '''WITH j AS(
WITH i AS (SELECT username,
        COUNT(rating_id) as rating_cnt
FROM ratings 
GROUP BY username
HAVING COUNT(rating_id) > 48
ORDER BY rating_cnt DESC)

SELECT r.username,
        COUNT(review_id) AS review_cnt
FROM i 
INNER JOIN reviews as r on i.username=r.username
GROUP BY r.username)

SELECT AVG (review_cnt)
FROM j'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

avg
0  24.0

<div class="alert alert-info"> <b>Комментарий студента:</b> Новый вывод: среднее количество обзоров от пользователей, которые поставили больше 48 оценок, равно 24.</div> 

## Общие выводы по проекту

1. Посчитайте, сколько книг вышло после 1 января 2000 года.

*После 1 января 2020 года (не включая эту дату) вышло 819 книг.*

2. Для каждой книги посчитайте количество обзоров и среднюю оценку.

*Для заданной базы с 1000 книг имеем средний рейтинг от 5.0 до 1.5 и количество обзоров от 7 до  0.*

3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.

*Больше всего книг толще 50 страниц выпустило издательство Penguin Books - 42 книги.*

4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.

*Новый вывод:  Автор с самой высокой средней оценкой книг при условии что количество оценок больше 50 - J.K. Rowling/Mary GrandPre.*

5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

*Cреднее количество обзоров от пользователей, которые поставили больше 48 оценок, равно 24.*